source/slang/slang-ast-serialize.h

```diff
@@ -7,13 +7,188 @@
+the mechanism to turn a pointer into a serializable ASTSerialIndex and vice versa. There are some special functions for turning string like
+types to and forth.
+
+The final mechanism is that of 'Arrays'. An array allows reading or writing a chunk of data associated with a ASTSerialIndex. The chunk of
+data *must* hold data that is serializable. If the array holds pointers - then the serialized array must hold ASTSerialIndices that
+represent those pointers. When reading back in they are converted back.
+
+Arrays are the escape hatch that allows for more complex types to serialize. Dictionaries for example are saved as a serial type that is
+two ASTSerialIndices one to a keys array and one to a values array.
+
+Note that writing has two phases, serializing out into an ASTSerialWriter, and then secondly writing out to a stream. 
+
+NodeBase Types
+==============
+
+The ASTSerialTypeInfo mechanism is generally for *fields* of NodeBase types. That for NodeBase derived types we use the C++ extractors
+field list to work out the native fields offsets and types. With this we can then calculate the layout for NodeBase types such that they
+follow the requirements for serialization - such as alignment and so forth.
+
+This information is held in the ASTSerialClasses, which for a given ASTNodeType gives an ASTSerialClassInfo, that specifies fields for
+just that type. Super types fields need to be serialized too, and this information can be found by using the ClassReflectInfo to find the
+super type.
+
+Reading
+=======
+
+Due to the care in writing reading is relatively simple. We can just take the contents of the file and put in memory, as long as in memory
+it has an alignment of at least MAX_ALIGNMENT. Then we can build up an entries table by stepping through the data and writing the pointer.
+
+The toNative functions take an ASTSerialReader - this allows the implementation to ask for pointers and arrays from other parts of the serialized
+data. It also allows for types to be lazily reconstructed if necessary.
+
+Lazy reconstruction may be useful in the future to partially reconstruct a sub part of the serialized data. In the current implementation, lazy
+evaluation is used on Strings. The m_objects array holds all of the recreated native 'objects'. Since the objects can be derived from different
+base classes the associated Entry will describe what it really is.
+
+For the String type, we initially store the object pointer as null. If a string is requested from that index, we see if the object pointer is null,
+if it is we have to construct the StringRepresentation that will be used.
+
+An extra wrinkle is that we allow accessing of a serialized String as a Name or a string or a UnownedSubString. Fortunately a Name just holds a string,
+and a Name remains in scope as long as it's NamePool does which is passed in.
```

Comment:
I'm not sure I fully follow this bit. Are you saying we serialize the `NamePool` as part of the AST, or that we will have a `NamePool` available when deserializing so that we can use it to construct `Name`s from strings as needed?

Reply:
The latter we have a NamePool to deserialize Names into but additionally (and perhaps less obviously) it is using the NamePool to scope UnownedStringSlices (for example as held in Token type in AST). That when a 'string' is accessed it can be accessed in 3 ways - a String, a Name and an UnownedStringSlice. So how it's stored may change depending on how it is accessed.

Reply:
> we have a NamePool to deserialize Names into but additionally (and perhaps less obviously) it is using the NamePool to scope UnownedStringSlices (for example as held in Token type in AST)

There might be an argument there to do some collapsing/unification between the `NamePool` and the string pool type(s) that you'd introduced as part of IR serialization. We could have a base layer that just handles deduplication and scoping the lifetime of strings, and a higher-level layer that simply wraps those deduplicated strings up as `Name`s (which mostly exist just so that AST-level code doesn't get confused and try to use plain strings as names, thus breaking our deduplication assumptions)